builtin_promise.go

```diff
@@ -0,0 +1,610 @@
+					_error.self._putProp("errors", r.newArrayValues(errors), true, false, true)
+					pcap.reject(_error)
+				}
+				return _undefined
+			}, nil, "", nil, 1)
+
+			remainingElementsCount++
+			r.invoke(nextPromise, "then", pcap.resolveObj, onRejected)
+		})
+		remainingElementsCount--
+		if remainingElementsCount == 0 {
+			_error := r.builtin_new(r.global.AggregateError, nil)
+			_error.self._putProp("errors", r.newArrayValues(errors), true, false, true)
+			pcap.reject(_error)
+		}
+	})
+	return pcap.promise
+}
+
+func (r *Runtime) promise_race(call FunctionCall) Value {
+	c := r.toObject(call.This)
+	pcap := r.newPromiseCapability(c)
+
+	pcap.try(r.vm.ctx, func() {
+		promiseResolve := r.toCallable(c.self.getStr("resolve", nil))
+		iter := r.getIterator(call.Argument(0), nil)
+		r.iterate(iter, func(nextValue Value) {
+			nextPromise := promiseResolve(FunctionCall{This: c, Arguments: []Value{nextValue}})
+			r.invoke(nextPromise, "then", pcap.resolveObj, pcap.rejectObj)
+		})
+	})
+	return pcap.promise
+}
+
+func (r *Runtime) promise_reject(call FunctionCall) Value {
+	pcap := r.newPromiseCapability(r.toObject(call.This))
+	pcap.reject(call.Argument(0))
+	return pcap.promise
+}
+
+func (r *Runtime) promise_resolve(call FunctionCall) Value {
+	return r.promiseResolve(r.toObject(call.This), call.Argument(0))
+}
+
+func (r *Runtime) createPromiseProto(val *Object) objectImpl {
+	o := newBaseObjectObj(val, r.global.ObjectPrototype, classObject)
+	o._putProp("constructor", r.global.Promise, true, false, true)
+
+	o._putProp("catch", r.newNativeFunc(r.promiseProto_catch, nil, "catch", nil, 1), true, false, true)
+	o._putProp("finally", r.newNativeFunc(r.promiseProto_finally, nil, "finally", nil, 1), true, false, true)
+	o._putProp("then", r.newNativeFunc(r.promiseProto_then, nil, "then", nil, 2), true, false, true)
+
+	o._putSym(SymToStringTag, valueProp(asciiString(classPromise), false, false, true))
+
+	return o
+}
+
+func (r *Runtime) createPromise(val *Object) objectImpl {
+	o := r.newNativeConstructOnly(val, r.builtin_newPromise, r.global.PromisePrototype, "Promise", 1)
+
+	o._putProp("all", r.newNativeFunc(r.promise_all, nil, "all", nil, 1), true, false, true)
+	o._putProp("allSettled", r.newNativeFunc(r.promise_allSettled, nil, "allSettled", nil, 1), true, false, true)
+	o._putProp("any", r.newNativeFunc(r.promise_any, nil, "any", nil, 1), true, false, true)
+	o._putProp("race", r.newNativeFunc(r.promise_race, nil, "race", nil, 1), true, false, true)
+	o._putProp("reject", r.newNativeFunc(r.promise_reject, nil, "reject", nil, 1), true, false, true)
+	o._putProp("resolve", r.newNativeFunc(r.promise_resolve, nil, "resolve", nil, 1), true, false, true)
+
+	r.putSpeciesReturnThis(o)
+
+	return o
+}
+
+func (r *Runtime) initPromise() {
+	r.global.PromisePrototype = r.newLazyObject(r.createPromiseProto)
+	r.global.Promise = r.newLazyObject(r.createPromise)
+
+	r.addToGlobal("Promise", r.global.Promise)
```

Comment:
Is this going to interfere with our own implementation in the baas repo?

Reply:
I tested this with most of our async tests and there was no issue, what I'll try to do is keep only goja's and see what happens. I will pull out the promise changes so that we can keep the ball rolling with the rest (originally I didn't wanna add it but it didn't seem to be a big change or breaking tests).